### ISSUE TITLE
nn.linear not support bfloat16

### ISSUE THREAD

**Issue by Nightbringers** [pytorch/pytorch#97872]
### 🐛 Describe the bug

from torch.distributed.fsdp import (
    FullyShardedDataParallel as FSDP,
    MixedPrecision,
    BackwardPrefetch,
    ShardingStrategy,
    FullStateDictConfig,
    StateDictType, CPUOffload,
)

class Model(nn.Module):
    def __init__(
        self,
    ):
        super().__init__()
    
        self.to_pred = nn.Linear(dim, 1, bias = False)

    def forward(
        self,
        input
    ):
       
        pred = self.to_pred(input)
        return pred

bfSixteen = MixedPrecision(
    param_dtype=torch.bfloat16,
    # Gradient communication precision.
    reduce_dtype=torch.bfloat16,
    # Buffer precision.
    buffer_dtype=torch.bfloat16,
)
model = Model()
model = FSDP(model,
                 auto_wrap_policy=opt_auto_wrap_policy,
                 mixed_precision=bfSixteen ,
     
                 device_id=torch.cuda.current_device())
res = model(input)

File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    pred = self.to_pred(pooled)
      File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
pred = self.to_pred(pooled)
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
    pred = self.to_pred(pooled)
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return forward_call(*input, **kwargs)
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
        return forward_call(*input, **kwargs)return forward_call(*input, **kwargs)

  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return forward_call(*input, **kwargs)
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return forward_call(*input, **kwargs)
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return forward_call(*input, **kwargs)
  File "/nfs/users/weisongwei/miniconda3/envs/tf2.9/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
    RuntimeErrorreturn F.linear(input, self.weight, self.bias)    : 
    return F.linear(input, self.weight, self.bias)expected scalar type Float but found BFloat16return F.linear(input, self.weight, self.bias)

    
RuntimeErrorreturn F.linear(input, self.weight, self.bias)    : RuntimeError
return F.linear(input, self.weight, self.bias)RuntimeErrorexpected scalar type Float but found BFloat16: 
: RuntimeError
expected scalar type Float but found BFloat16expected scalar type Float but found BFloat16: 
RuntimeError
expected scalar type Float but found BFloat16    : 

### Versions

torch2.0 and torch1.13  

when use fsdp wrap a model which have nn.Linear , if mixed_precision use bfloat16, it will be an error.

**Comment by awgu** [comment]
I ran the following, and I did not see any error.
```
"""
torchrun --standalone --nproc_per_node=2 test/test_lin_bf16.py
"""
import os

import torch
import torch.distributed as dist
import torch.nn as nn
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP, MixedPrecision
from torch.distributed.fsdp.wrap import ModuleWrapPolicy



class Model(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.to_pred = nn.Linear(5, 1, bias=False)

    def forward(self, inp: torch.Tensor) -> torch.Tensor:
        pred = self.to_pred(inp)
        return pred


def main():
    model = Model()
    # Either `auto_wrap_policy=None` or wrapping `nn.Linear` works
    auto_wrap_policy = ModuleWrapPolicy({nn.Linear})
    model = FSDP(
        model,
        auto_wrap_policy=auto_wrap_policy,
        mixed_precision=MixedPrecision(torch.bfloat16),
        device_id=torch.cuda.current_device(),
    )
    inp = torch.randn((2, 5), device="cuda")
    res = model(inp)


if __name__ == "__main__":
    dist.init_process_group(backend="nccl")
    gpu_id = int(os.environ["LOCAL_RANK"])
    device = f"cuda:{gpu_id}"
    torch.cuda.set_device(device)
    main()
```
I am not sure what your `opt_auto_wrap_policy` is, but I tried with no auto wrapping policy or one that only wraps `nn.Linear`, where both worked.

It would be great if you could provide an exact repro, or it might be more challenging for us to help.

**Comment by Nightbringers** [comment]
"""
torchrun --nnodes 1 --nproc_per_node 2  test/test_lin_bf16.py
"""
import os
import functools
import torch
import torch.distributed as dist
import torch.nn as nn
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP, MixedPrecision
# from torch.distributed.fsdp.wrap import ModuleWrapPolicy
from transformers import T5Tokenizer, BloomModel
from einops import rearrange, reduce, pack, unpack
from einops.layers.torch import Rearrange, Reduce
from torch.distributed.fsdp import (
    FullyShardedDataParallel as FSDP,
    MixedPrecision,
    BackwardPrefetch,
    ShardingStrategy,
    FullStateDictConfig,
    StateDictType, CPUOffload,
)
from torch.distributed.fsdp.wrap import (
    transformer_auto_wrap_policy,
    enable_wrap,
    wrap,
)
def exists(val):
    return val is not None
def masked_mean(seq, mask = None, dim = 1):
    if not exists(mask):
        return seq.mean(dim = dim)

    if seq.ndim == 3:
        mask = rearrange(mask, 'b n -> b n 1')

    masked_seq = seq.masked_fill(~mask, 0.)
    numer = masked_seq.sum(dim = dim)
    denom = mask.sum(dim = dim)

    masked_mean = numer / denom.clamp(min = 1e-3)
    masked_mean = masked_mean.masked_fill(denom == 0, 0.)
    return masked_mean
class Model(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.to_pred = nn.Linear(1024, 1, bias=False)
        self.model =  BloomModel.from_pretrained('bigscience/bloom-560m')

    def forward(self, inp: torch.Tensor,mask) -> torch.Tensor:
        
        out = self.model(inp)
        out = out['last_hidden_state']
        pooled = masked_mean(out, mask.bool(), dim = 1)
        pred = self.to_pred(pooled)
        return pred

def main():
    model = Model()
    local_rank = int(os.environ['LOCAL_RANK'])
    rank = int(os.environ['RANK'])
    world_size = int(os.environ['WORLD_SIZE'])
    dist.init_process_group(backend="nccl")
    # Either `auto_wrap_policy=None` or wrapping `nn.Linear` works
    # auto_wrap_policy = ModuleWrapPolicy({nn.Linear})
    torch.cuda.set_device(local_rank)
    model = FSDP(model,
                 mixed_precision=MixedPrecision(torch.bfloat16),
                 device_id=torch.cuda.current_device())
    inp = torch.tensor([[1,3,4,5,6]])
    mask = torch.tensor([[1,1,1,1,1]])
    res = model(inp.to(local_rank),mask.to(local_rank))
    print(res)


if __name__ == "__main__":
    main()


try this code
